### PR TITLE
CAF-2660: Audit Webservice Client Connection with HTTP/HTTPS proxy support

### PR DESCRIPTION
**Pull request by CJComac**

### CODE REVIEW

File: caf-audit/src/main/java/com/hpe/caf/auditing/webserviceclient/WebServiceClientAuditConnection.java [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+                    return new Proxy(Proxy.Type.HTTP, proxyInet);
+                }
+            }
+        }
+        LOG.debug(webServiceEndpointUrl.getHost() + " is included in the list of no_proxy hosts or there are no HTTP " +
+                "or HTTPS proxies set to base one upon.");
+        return null;
+    }
+
+    private String getWebserviceEndpointFullPath(String webserviceEndpoint) {
```
Comment (glucy) [inline caf-audit/src/main/java/com/hpe/caf/auditing/webserviceclient/WebServiceClientAuditConnection.java]: getWebserviceEndpointFullPath(String webserviceEndpoint)

getWeb**S**erviceEndpointFullPath(String web**S**erviceEndpoint)

File: caf-audit/src/main/java/com/hpe/caf/auditing/webserviceclient/WebServiceClientAuditEventBuilder.java [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+    private static void configureHttpUrlConnection(final HttpURLConnection httpUrlConnection,
+                                                   final int streamingModeLength) {
+        httpUrlConnection.setConnectTimeout(webServiceConnectionTimeout);
+        httpUrlConnection.setRequestProperty("Content-Type", "application/json; charset=UTF-8");
+        httpUrlConnection.setDoOutput(true);
+        //  The number of bytes to send is known; set fixed length streaming mode
+        httpUrlConnection.setFixedLengthStreamingMode(streamingModeLength);
+    }
+
+    private HttpURLConnection getWebserviceHttpEndpointUrlConnection(final int streamingModeLength) throws IOException {
```
Comment (glucy) [inline caf-audit/src/main/java/com/hpe/caf/auditing/webserviceclient/WebServiceClientAuditEventBuilder.java]: getWebserviceHttpEndpointUrlConnection

getWeb**S**erviceHttpEndpointUrlConnection

File: caf-audit/src/test/java/com/hpe/caf/auditing/webserviceclient/WebServiceClientAuditConnectionTest.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+import org.junit.Test;
+
+import java.lang.reflect.Field;
+import java.net.MalformedURLException;
+import java.net.UnknownHostException;
+import java.util.Map;
+
+public class WebServiceClientAuditConnectionTest {
+
+    String testWebserviceHttpsEndpoint = "https://testWsHost:8080/caf-audit-service/v1";
```
Comment (glucy) [inline caf-audit/src/test/java/com/hpe/caf/auditing/webserviceclient/WebServiceClientAuditConnectionTest.java]: testWebserviceHttpsEndpoint 

testWeb**S**erviceHttpsEndpoint

File: caf-audit/src/test/java/com/hpe/caf/auditing/webserviceclient/WebServiceClientAuditConnectionTest.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    @BeforeClass
+    public static void setup() {
+        // Test the Auditing library in webservice mode
+        System.setProperty("AUDIT_LIB_MODE", "webservice");
+    }
+
+    @Test(expected = MalformedURLException.class)
+    public void testWebserviceClientBadWebserviceEndpoint() throws Exception {
+
+        String invalidWebserviceEndpoint = "thisIsNotAValidEndpointUrl";
```
Comment (glucy) [inline caf-audit/src/test/java/com/hpe/caf/auditing/webserviceclient/WebServiceClientAuditConnectionTest.java]: invalidWebserviceEndpoint 

invalidWeb**S**erviceEndpoint

File: caf-audit-service-container/src/integration-test/java/com/hpe/caf/services/audit/api/AuditConnectionHelper.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    }
+
+    /**
+     * Returns an instance of an WebServiceClientAuditConnection (if environment variable AUDIT_LIB_MODE is set to
+     * 'webservice')
+     * @param webserviceEndpoint the webservice endpoint
+     * @return configured ElasticAuditConnection
+     * @throws ConfigurationException if there is an error with the ConfigurationSource
+     */
+    public static AuditConnection getWebserviceAuditConnection(String webserviceEndpoint)
```
Comment (glucy) [inline caf-audit-service-container/src/integration-test/java/com/hpe/caf/services/audit/api/AuditConnectionHelper.java]: getWebserviceAuditConnection(String webserviceEndpoint)

getWeb**S**erviceAuditConnection(String web**S**erviceEndpoint)

**Review by glucy — COMMENTED** [review]
Update these minor fixes then work with @mulhollandc  to implement the changes required from his most recent piece of work.

File: caf-audit-service-container/src/integration-test/java/com/hpe/caf/services/audit/api/AuditConnectionHelper.java (hunk 5 above)
Comment (CJComac) [reply caf-audit-service-container/src/integration-test/java/com/hpe/caf/services/audit/api/AuditConnectionHelper.java]: Adjusted in the latest commit

File: caf-audit/src/main/java/com/hpe/caf/auditing/webserviceclient/WebServiceClientAuditConnection.java (hunk 1 above)
Comment (CJComac) [reply caf-audit/src/main/java/com/hpe/caf/auditing/webserviceclient/WebServiceClientAuditConnection.java]: Adjusted in the latest commit

File: caf-audit/src/main/java/com/hpe/caf/auditing/webserviceclient/WebServiceClientAuditEventBuilder.java (hunk 2 above)
Comment (CJComac) [reply caf-audit/src/main/java/com/hpe/caf/auditing/webserviceclient/WebServiceClientAuditEventBuilder.java]: Adjusted in the latest commit

File: caf-audit/src/test/java/com/hpe/caf/auditing/webserviceclient/WebServiceClientAuditConnectionTest.java (hunk 3 above)
Comment (CJComac) [reply caf-audit/src/test/java/com/hpe/caf/auditing/webserviceclient/WebServiceClientAuditConnectionTest.java]: Adjusted in the latest commit

File: caf-audit/src/test/java/com/hpe/caf/auditing/webserviceclient/WebServiceClientAuditConnectionTest.java (hunk 4 above)
Comment (CJComac) [reply caf-audit/src/test/java/com/hpe/caf/auditing/webserviceclient/WebServiceClientAuditConnectionTest.java]: Adjusted in the latest commit

**Comment by glucy** [comment]
At this point the above changes are fine.  Please work with @mulhollandc to implement Connor's changes into your branch.

**Comment by CJComac** [comment]
I have modified the WebService client event builder to integrate with Connor's changes to the WebService's Audit Event input JSON schema

File: caf-audit-maven-plugin/caf-audit-maven-plugin.md [hunk 6]
```diff
@@ -49,12 +51,7 @@ provided next.
...
-description and an optional `ColumnName` element which identifies the database column to be used for storage in the event 
-that multiple audit events comprise of parameters with the same name. 
+`AuditedApplication` is the root element. `ApplicationId` identifies the name of the application that the audit event type is associated with. For each audit event defined, `TypeId` is a string identifier for the particular event (e.g. viewDocument) and `CategoryId` is a string identifier for the category of event. A list of parameter elements are then defined for each Audit Event. This includes the `Name` of the parameter, the `Type` (i.e. string, short, int, long, float, double, boolean or date) and the `Description`. The `IndexingHint` (i.e. fulltext or keyword) is optional and can be used to specify an indexing hint when storing audit event parameter data of `Type` string. The `Constraints` element is also optional and this can be used to specify minimum and/or maximum length constraints for audit event parameters of `Type` string. The `ColumnName` element too is optional which can be used to force the use of a particular database column when storing the audit data.
```
Comment (glucy) [inline caf-audit-maven-plugin/caf-audit-maven-plugin.md]: @mulhollandc and @CJComac 

"The `ColumnName` element too is optional which can be used to force the use of a particular database column when storing the audit data."

Is this still valid in today's implementation?  There is no concept of column's in Elasticsearch.

File: caf-audit-schema/README.md [hunk 7]
```diff
@@ -59,7 +60,7 @@ An example of an Audit Event Definition File is shown next:
...
+A list of parameter elements are then defined for each Audit Event. This includes the `Name` of the parameter, the `Type` (i.e. string, short, int, long, float, double, boolean or date) and the `Description`. The `IndexingHint` (i.e. fulltext or keyword) is optional and can be used to specify an indexing hint when storing audit event parameter data of `Type` string. Choose fulltext to support a full-text search on the string field after indexing, otherwise keyword to support a whole value search on the string only. The `Constraints` element is also optional and this can be used to specify minimum and/or maximum length constraints for audit event parameters of `Type` string. The `ColumnName` element too is optional which can be used to force the use of a particular database column when storing the audit data.
```
Comment (glucy) [inline caf-audit-schema/README.md]: As above?

File: pom.xml [hunk 8]
```diff
@@ -109,7 +109,7 @@
     </dependencyManagement>
 
     <modules>
-        <module>docs</module>
+        <!--<module>docs</module>-->
```
Comment (glucy) [inline pom.xml]: Needs removed before merging.

Comment (CJComac) [reply pom.xml]: Added back in the latest commit

**Comment by CJComac** [comment]
This should not be merged until "CAF-2693-support-ES-string-types" branch has been merged

**Comment by mulhollandc** [comment]
@glucy Changes relating to the indexing hint in commit 596567f look OK to me.

**Pull request merged**